src/PowerPing/Ping.cs

```diff
@@ -415,38 +399,54 @@ private PingResults SendICMP(PingAttributes attrs)
                     }
 
                     ICMP response;
-                    TimeSpan replyTime;
+                    EndPoint responseEP = iep;
+                    TimeSpan replyTime = TimeSpan.Zero;
                     do {
+                        // Cancel if requested
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        // Set receive timeout
+                        int remainingTimeout = (int)Math.Ceiling(attrs.Timeout - replyTime.TotalMilliseconds);
```

Comment:
I'm not sure if I like this, calling a math function like this I'm worried about the performance implications (maybe I'm overestimating .NETs overhead) and generally I'm not sure what the benefit of chopping up the timeout like this, is it so we can get quicker control-c? Just seems to be quite a bit of added math logic in general in this loop.

Reply:
If the socket timeout is long, let's say 5 seconds, there is no way to check the cancel signal if the receive operation is happening synchronously in the same thread.

My first workaround was to wrap Task.Run around it, which basically runs it in a separate thread and allows you to check the cancel signal in the first thread while waiting for the second thread to finish. But this adds a small latency to coordinate the threads. For example when I did some benchmarks pinging localhost, the response time went from 0.02ms to 0.07ms.

Chopping up the timeout allows us to check the cancel signal every 250ms without having to run the receive in a separate thread.